File: order-service/src/main/java/com/example/orderservice/application/logic/Result.java
package com.example.orderservice.application.logic;

import com.example.orderservice.domain.enums.ECustomErrorCode;

public class Result<T> {
  public final boolean isSuccess;
  public final boolean isFailure;
  private final ECustomErrorCode _errorCode;
  private final String _errorMessage;
  private final T _value;

  public Result(boolean isSuccess, T value, ECustomErrorCode errorCode,
                String errorMessage) {
    if (isSuccess && errorCode != null) {
      throw new IllegalArgumentException("InvalidOperation: A result cannot be "
                                         + "successful and contain an error");
    }
    if (!isSuccess && errorCode == null) {
      throw new IllegalArgumentException("InvalidOperation: A failing result "
                                         + "needs to contain an error message");
    }
    this.isSuccess = isSuccess;
    this.isFailure = !isSuccess;
    this._value = value;
    this._errorCode = errorCode;
    this._errorMessage = errorMessage;
  }

  public String toString() {
    if (this.isSuccess) {
      return this._value.toString();
    } else {
      return this._errorCode + " - " + this._errorMessage;
    }
  }

  public static <T> Result<T> ok(T value) {
    return new Result<T>(true, value, null, null);
  }

  public static <T> Result<T> fail(ECustomErrorCode errorCode,
                                   String errorMessage) {
    return new Result<T>(false, null, errorCode, errorMessage);
  }

  public ECustomErrorCode getErrorCode() {
    if (this.isSuccess) {
      throw new IllegalStateException("InvalidOperation: Can't get the error "
                                      + "code from a successful result");
    }
    return this._errorCode;
  }

  public String getErrorMessage() {
    if (this.isSuccess) {
      throw new IllegalStateException("InvalidOperation: Can't get the error "
                                      + "message from a successful result");
    }
    return this._errorMessage;
  }

  public T get_value() {
    if (this.isFailure) {
      throw new IllegalStateException("InvalidOperation: Can't get the value "
                                      + "from a failed result");
    }
    return this._value;
  }
}

File: inventory-service/src/application/logic/result.go
package logic

type Result[T any] struct {
	IsSuccess     bool
	IsFailure     bool
	_value        T
	_errorMessage string
}

func OkResult[T any](value T) *Result[T] {
	return &Result[T]{true, false, value, ""}
}

func FailedResult[T any](errorMessage string) *Result[T] {
	var t T
	return &Result[T]{false, true, t, errorMessage}
}

func (r *Result[T]) GetValue() T {
	if r.IsFailure {
		panic("Cannot get value from a failed result")
	}
	return r._value
}

func (r *Result[T]) GetErrorMessage() string {
	if r.IsSuccess {
		panic("Cannot get error message from a successful result")
	}
	return r._errorMessage
}
